csharp/src/Ice/ObjectAdapter.cs

```diff
@@ -1193,6 +1068,132 @@ private void UpdateLocatorRegistry(LocatorInfo? locatorInfo, IObjectPrx? proxy)
+                // adapter to allow to user code to retry activating the adapter later.
+                lock (_mutex)
+                {
+                    _state = State.Uninitialized;
+                }
+                throw;
+            }
+
+            if ((Communicator.GetPropertyAsBool("Ice.PrintAdapterReady") ?? false) && Name.Length > 0)
+            {
+                Console.Out.WriteLine($"{Name} ready");
+            }
+
+            lock (_mutex)
+            {
+                Debug.Assert(_state == State.Activating);
+                foreach (IncomingConnectionFactory factory in _incomingConnectionFactories)
+                {
+                    factory.Activate();
+                }
+                _state = State.Active;
+            }
+        }
+
+        private async Task PerformDeactivateAsync()
+        {
+            // Wait for activation to complete. This is necessary avoid out of order locator updates.
+            if (_activateTask != null)
+            {
+                await _activateTask.ConfigureAwait(false);
+            }
+
+            lock (_mutex)
+            {
+                Debug.Assert(_state <= State.Deactivating);
+                _state = State.Deactivating;
+            }
+
+            // Note: the router/locator infos and incoming connection factory list are immutable at this point.
+            try
+            {
+                if (_routerInfo != null)
+                {
+                    // Remove entry from the router manager.
+                    Communicator.EraseRouterInfo(_routerInfo.Router);
+
+                    // Clear this object adapter with the router.
+                    _routerInfo.Adapter = null;
+                }
+
+                await UpdateLocatorRegistryAsync(_locatorInfo, null);
+            }
+            catch
+            {
+                // We can't throw exceptions in deactivate so we ignore failures to update the locator registry.
+            }
+
+            lock (_mutex)
+            {
+                if (_directCount > 0)
+                {
+                    Debug.Assert(_directCountSemaphore == null);
+                    _directCountSemaphore = new SemaphoreSlim(_directCount);
+                }
+            }
+
+            if (_directCountSemaphore != null)
+            {
+                await _directCountSemaphore.WaitAsync().ConfigureAwait(false);
+            }
+
+            foreach (IncomingConnectionFactory factory in _incomingConnectionFactories)
+            {
+                _ = factory.DestroyAsync();
+            }
+
+            Communicator.OutgoingConnectionFactory().RemoveAdapter(this);
+
+            lock (_mutex)
+            {
+                Debug.Assert(_state == State.Deactivating);
+                _state = State.Deactivated;
+            }
+        }
+
+        private async Task PerformDisposeAsync()
```

Comment:
Shouldn't this method await on DestroyAsync for each incoming connection factory?

Reply:
Shouldn't we just await incoming factory destruction in PerformDeactivateAsync?

Reply:
What's the point of keeping DeactivateAsync actually? We could just keep DisposeAsync which indeed awaits on the factories destruction.

Reply:
Right we can keep just DisposeAsync

Reply:
Removed DeactivateAsync